File: backend/verification-service/src/controllers/verificationController.ts
import { Request, Response, NextFunction } from 'express';
import { VerificationService } from '@services/verificationService';
import logger from '@utils/logger';

/**
 * Controller for verification endpoints
 */
export class VerificationController {
  /**
   * POST /api/verify - Verify certificate by ID or code
   */
  static async verify(req: Request, res: Response, next: NextFunction) {
    try {
      const { certificateId, verificationCode } = req.body;
      const verifierIp = req.ip;

      let result;

      if (certificateId) {
        result = await VerificationService.verifyById(certificateId, verifierIp);
      } else if (verificationCode) {
        result = await VerificationService.verifyByCode(verificationCode, verifierIp);
      } else {
        return res.status(400).json({
          success: false,
          error: {
            code: 'INVALID_REQUEST',
            message: 'Either certificateId or verificationCode must be provided',
            timestamp: new Date().toISOString(),
          },
        });
      }

      return res.status(200).json({
        success: true,
        data: result,
        message: result.valid ? 'Certificate verified successfully' : 'Certificate verification failed',
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/verify/:certificateId - Quick verify by ID
   */
  static async verifyById(req: Request, res: Response, next: NextFunction) {
    try {
      const { certificateId } = req.params;
      const verifierIp = req.ip;

      const result = await VerificationService.verifyById(certificateId, verifierIp);

      return res.status(200).json({
        success: true,
        data: result,
        message: result.valid ? 'Certificate verified successfully' : 'Certificate verification failed',
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/verify/code/:verificationCode - Quick verify by code
   */
  static async verifyByCode(req: Request, res: Response, next: NextFunction) {
    try {
      const { verificationCode } = req.params;
      const verifierIp = req.ip;

      const result = await VerificationService.verifyByCode(verificationCode, verifierIp);

      return res.status(200).json({
        success: true,
        data: result,
        message: result.valid ? 'Certificate verified successfully' : 'Certificate verification failed',
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * POST /api/verify/bulk - Bulk verification
   */
  static async verifyBulk(req: Request, res: Response, next: NextFunction) {
    try {
      const { certificates } = req.body;
      const verifierIp = req.ip;

      const results = await VerificationService.verifyBulk(certificates, verifierIp);

      const totalRequested = results.length;
      const validCertificates = results.filter((r) => r.valid).length;
      const invalidCertificates = totalRequested - validCertificates;

      return res.status(200).json({
        success: true,
        data: {
          totalRequested,
          validCertificates,
          invalidCertificates,
          results,
        },
        message: `Bulk verification completed. ${validCertificates}/${totalRequested} certificates are valid.`,
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * POST /api/verify/signature - Verify digital signature
   */
  static async verifySignature(req: Request, res: Response, next: NextFunction) {
    try {
      const { certificateHash, digitalSignature, universityId } = req.body;
      const verifierIp = req.ip;

      const result = await VerificationService.verifySignature(
        certificateHash,
        digitalSignature,
        universityId,
        verifierIp
      );

      return res.status(200).json({
        success: true,
        data: result,
        message: result.valid ? 'Digital signature verified successfully' : 'Digital signature verification failed',
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/verify/statistics - Get verification statistics (admin)
   */
  static async getStatistics(req: Request, res: Response, next: NextFunction) {
    try {
      const days = parseInt(req.query.days as string) || 30;

      const statistics = await VerificationService.getStatistics(days);

      return res.status(200).json({
        success: true,
        data: {
          period: `Last ${days} days`,
          statistics,
        },
        message: 'Statistics retrieved successfully',
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * GET /api/verify/history/:certificateId - Get verification history
   */
  static async getHistory(req: Request, res: Response, next: NextFunction) {
    try {
      const { certificateId } = req.params;
      const limit = parseInt(req.query.limit as string) || 10;

      const history = await VerificationService.getVerificationHistory(certificateId, limit);

      return res.status(200).json({
        success: true,
        data: {
          certificateId,
          totalRecords: history.length,
          history,
        },
        message: 'Verification history retrieved successfully',
      });
    } catch (error) {
      next(error);
    }
  }
}

File: backend/verification-service/src/main/java/com/certverify/verification/controller/VerificationController.java
package com.certverify.verification.controller;

import com.certverify.verification.dto.BulkVerificationRequest;
import com.certverify.verification.dto.VerificationRequest;
import com.certverify.verification.dto.VerificationResponse;
import com.certverify.verification.model.VerificationResult;
import com.certverify.verification.service.VerificationService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Pattern;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/verify")
@RequiredArgsConstructor
@Validated
@Tag(name = "Verification", description = "Certificate verification endpoints")
public class VerificationController {

    private final VerificationService verificationService;

    /**
     * Verify certificate by ID or code
     */
    @PostMapping
    @Operation(summary = "Verify certificate", description = "Verify certificate by ID or verification code")
    public ResponseEntity<VerificationResponse> verify(@Valid @RequestBody VerificationRequest request) {

        VerificationResult result;

        if (request.getCertificateId() != null && !request.getCertificateId().isEmpty()) {
            result = verificationService.verifyById(request.getCertificateId());
        } else if (request.getVerificationCode() != null && !request.getVerificationCode().isEmpty()) {
            result = verificationService.verifyByCode(request.getVerificationCode());
        } else {
            return ResponseEntity.badRequest().body(
                    VerificationResponse.builder()
                            .success(false)
                            .message("Either certificateId or verificationCode must be provided")
                            .build()
            );
        }

        String message = result.getValid()
                ? "Certificate verified successfully"
                : "Certificate verification failed";

        return ResponseEntity.ok(
                VerificationResponse.builder()
                        .success(true)
                        .data(result)
                        .message(message)
                        .build()
        );
    }

    /**
     * Verify certificate by ID (GET)
     */
    @GetMapping("/{certificateId}")
    @Operation(summary = "Verify by ID", description = "Quick verification using certificate ID")
    public ResponseEntity<VerificationResponse> verifyById(
            @PathVariable
            @Pattern(regexp = "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
            String certificateId) {

        VerificationResult result = verificationService.verifyById(certificateId);

        String message = result.getValid()
                ? "Certificate verified successfully"
                : "Certificate verification failed";

        return ResponseEntity.ok(
                VerificationResponse.builder()
                        .success(true)
                        .data(result)
                        .message(message)
                        .build()
        );
    }

    /**
     * Verify certificate by code (GET)
     */
    @GetMapping("/code/{verificationCode}")
    @Operation(summary = "Verify by code", description = "Quick verification using verification code")
    public ResponseEntity<VerificationResponse> verifyByCode(
            @PathVariable
            @Pattern(regexp = "^[A-Z0-9]{6,8}$")
            String verificationCode) {

        VerificationResult result = verificationService.verifyByCode(verificationCode);

        String message = result.getValid()
                ? "Certificate verified successfully"
                : "Certificate verification failed";

        return ResponseEntity.ok(
                VerificationResponse.builder()
                        .success(true)
                        .data(result)
                        .message(message)
                        .build()
        );
    }

    /**
     * Bulk verification
     */
    @PostMapping("/bulk")
    @Operation(summary = "Bulk verification", description = "Verify multiple certificates at once")
    public ResponseEntity<Map<String, Object>> bulkVerify(@Valid @RequestBody BulkVerificationRequest request) {

        List<Map<String, Object>> results = new ArrayList<>();
        int validCount = 0;
        int invalidCount = 0;

        for (VerificationRequest certRequest : request.getCertificates()) {
            try {
                VerificationResult result;

                if (certRequest.getCertificateId() != null) {
                    result = verificationService.verifyById(certRequest.getCertificateId());
                } else {
                    result = verificationService.verifyByCode(certRequest.getVerificationCode());
                }

                if (result.getValid()) {
                    validCount++;
                } else {
                    invalidCount++;
                }

                Map<String, Object> resultMap = new HashMap<>();
                resultMap.put("certificateId", certRequest.getCertificateId());
                resultMap.put("verificationCode", certRequest.getVerificationCode());
                resultMap.put("valid", result.getValid());
                resultMap.put("reason", result.getReason());

                results.add(resultMap);

            } catch (Exception e) {
                invalidCount++;
                Map<String, Object> errorResult = new HashMap<>();
                errorResult.put("certificateId", certRequest.getCertificateId());
                errorResult.put("verificationCode", certRequest.getVerificationCode());
                errorResult.put("valid", false);
                errorResult.put("reason", "Verification failed due to internal error");
                results.add(errorResult);
            }
        }

        Map<String, Object> response = new HashMap<>();
        response.put("success", true);

        Map<String, Object> data = new HashMap<>();
        data.put("totalRequested", request.getCertificates().size());
        data.put("validCertificates", validCount);
        data.put("invalidCertificates", invalidCount);
        data.put("results", results);

        response.put("data", data);
        response.put("message", String.format("Bulk verification completed. %d/%d certificates are valid.",
                validCount, request.getCertificates().size()));

        return ResponseEntity.ok(response);
    }
}